openstf/model_selection/model_selection.py

```diff
@@ -1,87 +1,254 @@
             fraction of test data.
         validation_fraction (float): Number between 0 and 1 that indicates the
             desired fraction of validation data.
-        backtest (bool): Indicates if data is intended for a back test.
-        period_sampling (bool): Indicates if validation data must be sampled as
-            periods.
-        period_timedelta (datetime.timedelta): Duration of the periods that will
-            be sampled as validation data. Only used for period_sampling=True.
+        back_test (bool): Indicates if data is intended for a back test.
+        stratification (bool): Indicates if validation data must be sampled as
+            periods, using stratification
 
     Returns:
-        Tuple with train data, validation data and test data:
-            [0] (pandas.DataFrame): Train data
-            [1] (pandas.DataFrame): Validation data
-            [2] (pandas.DataFrame): Test data
+        train_data (pandas.DataFrame): Train data.
+        validation_data (pandas.DataFrame): Validation data.
+        test_data (pandas.DataFrame): Test data.
+
     """
-    MIN_TRAIN_FRACTION = 0.5
-    logger = structlog.get_logger(__name__)
 
     # Check input
-    train_fraction = 1 - (test_fraction + validation_fraction)
+    data_["timestamp"] = data_.index
+    if "Horizon" in data_.columns:
+        data = data_[data_["Horizon"] == 47]
+    else:
+        data = data_
+        stratification = False
 
+    data["idx_num"] = range(0, len(data))
+
+    train_fraction = 1 - (test_fraction + validation_fraction)
+    train_val_fraction = train_fraction + validation_fraction
     if train_fraction < 0:
         raise ValueError(
             "Test ({test_fraction}) and validation fraction ({validation_fraction}) too high."
         )
 
-    if train_fraction < MIN_TRAIN_FRACTION:
-        # TODO no action if above threshold? Which settings are meant here?
-        logger.warning("Current settings only allow for 50% train data")
-
     # Get start date from the index
     start_date = data.index.min().to_pydatetime()
 
     # Calculate total of quarter hours (PTU's) in input data
     number_indices = len(data.index.unique())  # Total number of unique timepoints
     delta = (
         data.index.unique().sort_values()[1] - data.index.unique().sort_values()[0]
-    )  # Delta t, assumed to be constant troughout DataFrame
+    )  # Delta t, assumed to be constant throughout DataFrame
     delta = timedelta(
         seconds=delta.seconds
     )  # Convert from pandas timedelta to original python timedelta
 
+    # Identify the peaks and list them
+    data.loc[:, "int_days"] = (
+        pd.to_datetime(data.index, format="%Y%m%d")
+        .strftime("%Y%m%d")
+        .astype(int)
+        .tolist()
+    )
+    data.loc[:, "days"] = np.where(
+        (data.int_days.diff().fillna(0).astype(int)[: (len(data))] > 0),
+        1,
+        data.int_days.diff().fillna(0).astype(int)[: (len(data))],
+    ).cumsum()
```

Comment:
Looks quite complex. I suppose these lines of code have something to do with coupling the rownumber af the dataframe to a specific date? I think we can make it a bit more clean using the index.date() functionality. Let's have a look together